File: lib/features/prontuario/domain/entities/treatment_plan.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'treatment_plan.freezed.dart';
part 'treatment_plan.g.dart';

@freezed
class TreatmentPlan with _$TreatmentPlan {
  const factory TreatmentPlan({
    required String id,
    required String patientId,
    required String description,
    required List<TreatmentItem> items,
    required String createdByUserId,
    required TreatmentPlanStatus status,
    required DateTime createdAt,
    DateTime? updatedAt,
  }) = _TreatmentPlan;

  factory TreatmentPlan.fromJson(Map<String, dynamic> json) => _$TreatmentPlanFromJson(json);

  Map<String, dynamic> toJson() => _$$TreatmentPlanImplToJson(this as _$TreatmentPlanImpl);
}

@freezed
class TreatmentItem with _$TreatmentItem {
  const factory TreatmentItem({
    required String id,
    required String procedureId,
    required String procedureName,
    int? toothNumber,
    String? observation,
    required TreatmentItemStatus status,
  }) = _TreatmentItem;

  factory TreatmentItem.fromJson(Map<String, dynamic> json) => _$TreatmentItemFromJson(json);

  Map<String, dynamic> toJson() => _$$TreatmentItemImplToJson(this as _$TreatmentItemImpl);
}

enum TreatmentPlanStatus {
  draft,
  approved,
  inProgress,
  completed,
  cancelled,
}

enum TreatmentItemStatus {
  pending,
  inProgress,
  completed,
  cancelled,
}

File: backend/DentalClinic.Core/Domain/Entities/TreatmentPlan.cs
using System;
using System.Collections.Generic;

namespace DentalClinic.Core.Domain.Entities;

/// <summary>
/// Representa o Plano de Tratamento proposto para um paciente.
/// Contém uma lista de procedimentos a serem realizados.
/// </summary>
public sealed class TreatmentPlan : Entity
{
    public Guid PatientId { get; private set; }
    public Patient Patient { get; private set; } = null!;

    public string Description { get; private set; } = string.Empty;

    public List<TreatmentItem> Items { get; private set; } = new();

    public Guid CreatedByUserId { get; private set; }
    public User CreatedBy { get; private set; } = null!;

    public TreatmentPlanStatus Status { get; private set; }

    private TreatmentPlan() { }

    public static TreatmentPlan Create(Guid patientId, Guid createdByUserId, string description)
    {
        return new TreatmentPlan
        {
            PatientId = patientId,
            CreatedByUserId = createdByUserId,
            Description = description,
            Status = TreatmentPlanStatus.Draft
        };
    }

    public void AddItem(Guid procedureId, string procedureName, decimal value, int? toothNumber)
    {
        Items.Add(TreatmentItem.Create(Id, procedureId, procedureName, value, toothNumber));
        UpdatedAt = DateTime.UtcNow;
    }

    public void Approve()
    {
        Status = TreatmentPlanStatus.Approved;
        UpdatedAt = DateTime.UtcNow;
    }

    public void Complete()
    {
        Status = TreatmentPlanStatus.Completed;
        UpdatedAt = DateTime.UtcNow;
    }
}

public enum TreatmentPlanStatus
{
    Draft = 1,
    Approved = 2,
    InProgress = 3,
    Completed = 4,
    Cancelled = 5
}
